File: attendance/attendance_api.go
package main

import (
	"attendance/config"
	"database/sql"
	// 	_ "github.com/go-sql-driver/mysql"
	"go.elastic.co/apm/module/apmgin/v2"
	"go.elastic.co/apm/module/apmsql/v2"
	_ "go.elastic.co/apm/module/apmsql/v2/mysql"

	"github.com/gin-contrib/cors"
	"github.com/gin-gonic/gin"
	"github.com/sirupsen/logrus"

	"net/http"
	"os"
)

const (
	dbDriver = "mysql"
)

var (
	configFile = os.Getenv("CONFIG_FILE")
)

// AttendanceInfo struct will be the data structure for employee's attendance
type AttendanceInfo struct {
	ID     int    `json:"id"`
	Date   string `json:"date"`
	Status string `json:"status"`
}

func main() {
	conf, err := config.ParseFile(configFile)
	logrus.SetFormatter(&logrus.JSONFormatter{})
	if err != nil {
		logrus.Errorf("Unable to parse configuration file for attendance: %v", err)
	}
	logrus.Infof("Running employee-attendance in webserver mode")
	logrus.Infof("employee-attendance api is listening on port: %v", conf.Attendance.APIPort)
	logrus.Infof("Endpoint is available now - http://0.0.0.0:%v/create", conf.Attendance.APIPort)
	router := gin.Default()
	router.Use(apmgin.Middleware(router))
	config := cors.DefaultConfig()
	config.AllowOrigins = []string{"*"}
	router.Use(cors.New(config))
	router.POST("/attendance/create", pushAttendanceData)
	router.GET("/attendance/search", fetchAttendanceData)
	router.GET("attendance/healthz", healthCheckMySQL)
	router.Run(":" + conf.Attendance.APIPort)
}

func initDBConnection() (*sql.DB, error) {
	conf, err := config.ParseFile(configFile)
	if err != nil {
		logrus.Errorf("Unable to parse configuration file for attendance: %v", err)
	}
	db, err := apmsql.Open(dbDriver, conf.MySQL.Username+":"+conf.MySQL.Password+"@tcp("+conf.MySQL.Host+")/"+conf.MySQL.DBName)
	if err != nil {
		return db, err
	}
	return db, nil
}

func pushAttendanceData(c *gin.Context) {
	conf, err := config.ParseFile(configFile)
	if err != nil {
		logrus.Errorf("Unable to parse configuration file for attendance: %v", err)
	}
	db, err := initDBConnection()
	tx, err := db.BeginTx(c.Request.Context(), nil)
	if err != nil {
		logrus.Errorf("Error while creating sql connection for pushing attendance data: %v", err)
	}
	_, err = tx.ExecContext(c.Request.Context(), "USE "+conf.MySQL.DBName)
	if err != nil {
		logrus.Errorf("Not able to use database: %v", err)
	}
	_, err = tx.ExecContext(c.Request.Context(), "CREATE TABLE IF NOT EXISTS Employee ( id int(6) NOT NULL, status varchar(50) NOT NULL, date varchar(50), PRIMARY KEY (id) )")
	if err != nil {
		logrus.Errorf("Error while creating Table: %v", err)
	}
	defer db.Close()

	var request AttendanceInfo
	if err := c.BindJSON(&request); err != nil {
		errorResponse(c, http.StatusBadRequest, "Malformed request body")
		logrus.Errorf("Error parsing the request body in JSON: %v", err)
		return
	}
	insForm, err := tx.Prepare("INSERT INTO Employee(id, status, date) VALUES(?,?,?)")

	if err != nil {
		logrus.Errorf("Cannot create db insertion command: %v", err)
	}

	insForm.ExecContext(c.Request.Context(), request.ID, request.Status, request.Date)
	if err := tx.Commit(); err != nil {
		logrus.Errorf("Cannot create db insertion command: %v", err)
	}
	defer db.Close()
	logrus.Infof("Successfully pushed employee's attendance information")
}

func fetchAttendanceData(c *gin.Context) {
	db, err := initDBConnection()
	tx, err := db.BeginTx(c.Request.Context(), nil)
	if err != nil {
		logrus.Errorf("Error while creating sql connection for fetching attendance data: %v", err)
	}
	selDB, err := tx.QueryContext(c.Request.Context(), "SELECT * FROM Employee ORDER BY id DESC")

	var attendanceInfo []AttendanceInfo

	for selDB.Next() {
		var id int
		var status, date string
		err = selDB.Scan(&id, &status, &date)
		if err != nil {
			logrus.Errorf("Error while scanning data: %v", err)
		}

		attendanceInfo = append(attendanceInfo, AttendanceInfo{
			ID:     id,
			Status: status,
			Date:   date,
		})
	}
	if err := tx.Commit(); err != nil {
		logrus.Errorf("Cannot create db fetch command: %v", err)
	}
	c.JSON(http.StatusOK, attendanceInfo)
}

func healthCheckMySQL(c *gin.Context) {
	db, err := initDBConnection()
	if err != nil {
		logrus.Errorf("Error while creating sql connection for fetching attendance data: %v", err)
	}

	err = db.PingContext(c.Request.Context())
	if err != nil {
		logrus.Errorf("Unable to communicate with MySQL database: %v", err)
		errorResponse(c, http.StatusBadRequest, "MySQL connection is not up")
		return
	}
	defer db.Close()

	c.JSON(http.StatusOK, gin.H{
		"status":   "up",
		"database": "MySQL",
		"message":  "MySQL is running",
	})
}

func errorResponse(c *gin.Context, code int, err string) {
	c.JSON(code, gin.H{
		"error": err,
	})
}

File: attendance/attendance_api.py
#!/usr/local/bin/python
# pylint: disable=no-member, bare-except, line-too-long
"""
Attendance application is part of OT-Microservices.
This application stores and pull attendance data from MySQL
Available Endpoints:-
    - /attendance/create
    - /attendance/search
    - /attendance/healthz
"""

import json
import os
import logging
import yaml
from flask import Flask, jsonify, request
import mysql.connector
from elasticapm.contrib.flask import ElasticAPM
from elasticapm.handlers.logging import LoggingHandler

ELASTIC_APM_SERVER_URL = os.getenv('ELASTIC_APM_SERVER_URL', 'http://localhost:8200')
ELASTIC_APM_SERVICE_NAME = os.getenv('ELASTIC_APM_SERVICE_NAME', 'attendance')
CONFIG_FILE = os.getenv('CONFIG_FILE', '/app/config/config.yaml')

app = Flask(__name__)

def read_config():
    """Method for reading the configuration file for attendance"""
    with open(CONFIG_FILE, 'r', encoding="utf-8") as config_file:
        yaml_values = yaml.load(config_file, Loader=yaml.FullLoader)
    return yaml_values

config_properties = read_config()

apm = ElasticAPM(app,
                server_url=ELASTIC_APM_SERVER_URL,
                service_name=ELASTIC_APM_SERVICE_NAME,
                logging=True)

app.config['MYSQL_HOST'] = config_properties['mysql']['host']
app.config['MYSQL_USERNAME'] = config_properties['mysql']['username']
app.config['MYSQL_PASSWORD'] = config_properties['mysql']['password']
app.config['MYSQL_DATABASE'] = config_properties['mysql']['db_name']

@app.route("/attendance/healthz", methods=['GET'])
def check_health():
    """Method for checking the health of MySQL"""
    try:
        connection = create_mysql_client()
        connection.ping()
        return jsonify(mysql="up", description="MySQL is healthy"), 200
    except:
        app.logger.error("Unable to make mysql connection", exc_info=True)
        return jsonify(mysql="down", description="MySQL is not healthy"), 400

@app.route("/attendance/create", methods=['POST'])
def push_attendance_data():
    """For pushing attendance data inside MySQL Database"""
    record = json.loads(request.data)
    try:
        connection = create_mysql_client()
        cursor = connection.cursor()
        sql = "CREATE TABLE IF NOT EXISTS Employee ( id int(6) NOT NULL, status varchar(50) NOT NULL, date varchar(50), PRIMARY KEY (id) )"
        cursor.execute(sql)
    except:
        app.logger.error("Unable to create table in MySQL", exc_info=True)
        return 400

    try:
        cursor = connection.cursor()
        sql = "INSERT INTO `Employee` (`id`, `status`, `date`) VALUES (%s, %s, %s)"
        cursor.execute(sql, (record['id'], record['status'], record['date']))
        app.logger.info("Successfully pushed attendance data in MySQL", exc_info=True)
        return jsonify(message="Successfully uploaded the attendance data"), 200
    except:
        app.logger.error("Unable to push attendance data in MySQL", exc_info=True)
        return jsonify(message="Error in pushing attendance data"), 400

@app.route("/attendance/search", methods=['GET'])
def fetch_attendance_data():
    """For pulling attendance data from MySQL Database"""
    try:
        connection = create_mysql_client()
        cursor = connection.cursor()
        sql = "SELECT * FROM Employee ORDER BY id DESC"
        cursor.execute(sql)
        results = cursor.fetchall()
        complete_data = []
        data = {}
        for result in results:
            data["id"] = result[0]
            data["status"] = result[1]
            data["date"] = result[2]
            complete_data.append(data.copy())
        app.logger.info("Successfully pulled attendance data from MySQL", exc_info=True)
        return jsonify(complete_data), 200
    except:
        app.logger.error("Unable to pull attendance data from MySQL", exc_info=True)
        return jsonify(message="Error while pulling data for attendance"), 200

def create_mysql_client():
    """For creating the client connection with MySQL"""
    connection = mysql.connector.connect(
        host=app.config['MYSQL_HOST'],
        user=app.config['MYSQL_USERNAME'],
        passwd=app.config['MYSQL_PASSWORD'],
        database=app.config['MYSQL_DATABASE']
    )
    return connection

if __name__ == "__main__":
    handler = LoggingHandler(client=apm.client)
    console_handler = logging.StreamHandler()
    app.logger.addHandler(handler)
    app.logger.addHandler(console_handler)
    app.run()
